test/helpers/value_hook.dart

```diff
@@ -49,48 +223,190 @@ class VHook<T> {
         skip: skip,
       );
 
-  /// Get awaitable that completes when the contained value is non-null.
+  /// Wait for the value to conform to [matcher].
+  ///
+  /// The optional [timeout] specifies the maximum time to wait for the match
+  /// before a [TimeoutException] is thrown.
+  ///
+  /// By default all intermediate states are checked, to only check after
+  /// completion enable [onlyOnCompletion].
+  ///
+  /// This method uses 'expect' from the 'test' package.
   ///
-  /// Note: The initial value of the VHook must have been null, and the final
-  /// value non-null.
+  /// [matcher] is used to test for the expected value
+  /// e.g. `equals(true)`
+  Future<void> expectAsync(
+    test.Matcher matcher, {
+    String? reason,
+    dynamic skip,
+    Duration? timeout,
+    bool onlyOnCompletion = false,
+  }) async {
+    if (onlyOnCompletion) {
+      // Only check after completion
+      await awaitCompletion(timeout);
+    } else {
+      // Check all intermediate states
+      try {
+        await awaitValue(
+          timeout: timeout,
+          condition: (T val) {
+            try {
+              test.expect(
+                _value,
+                matcher,
+                reason: reason,
+                skip: skip,
+              );
+
+              return true;
+            } on test.TestFailure {
+              return false;
+            }
+          },
+        );
+      } on TimeoutException {
+        // We catch TimeoutException to show the results of
+        // matching the most current value (the [TestFailure])
```

Comment:
Should this block be empty?

Reply:
Yes, but it is inconsistent with the docs. We have two options here:

1. After timeout throw a `TimeoutException`
2. After timeout use `test.expect`, which we know will throw (as we would have returned before if it wouldn’t) to show the user that expected and current state (which I would consider more useful)

The doc comment currently suggests 1. but the actual behaviour is 2..

Reply:
I'd throw a timeout exception on timeout. You know the keep things simple stuff ;)